src/main/java/tc/oc/pgm/map/NameCacheUtil.java

```diff
@@ -0,0 +1,122 @@
+import java.io.InputStreamReader;
+import java.net.HttpURLConnection;
+import java.net.URL;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map.Entry;
+import java.util.UUID;
+import org.bukkit.Bukkit;
+import tc.oc.pgm.api.PGM;
+
+public class NameCacheUtil {
+
+  private static final Gson GSON = new Gson();
+
+  private static List<UUID> unresolved = new ArrayList<UUID>();
+  private static HashMap<UUID, CachedPlayer> cache = new HashMap<UUID, CachedPlayer>();
+
+  public static boolean isUUIDCached(UUID uuid) {
+    return cache.containsKey(uuid);
+  }
+
+  public static CachedPlayer getCachedPlayer(UUID uuid) {
+    return cache.get(uuid);
+  }
+
+  public static void addUUID(UUID uuid) {
+    if (!cache.containsKey(uuid)) {
+      unresolved.add(uuid);
+    }
+  }
+
+  private static String resolveName(UUID uuid) throws IOException {
+    HttpURLConnection con =
+        (HttpURLConnection)
+            new URL("https://api.ashcon.app/mojang/v2/user/" + uuid.toString()).openConnection();
+    con.setRequestMethod("GET");
+    con.setRequestProperty("Accept", "application/json");
+    con.setConnectTimeout(5000);
+    con.setReadTimeout(5000);
+    BufferedReader br =
+        new BufferedReader(new InputStreamReader(con.getInputStream(), StandardCharsets.UTF_8));
+    StringBuilder response = new StringBuilder();
+    String line;
+    while ((line = br.readLine()) != null) {
+      response.append(line.trim());
+    }
+    br.close();
+    return GSON.fromJson(response.toString(), JsonObject.class).get("username").getAsString();
+  }
+
+  public static void resolveAll() {
+    Bukkit.getScheduler()
+        .runTaskAsynchronously(
+            PGM.get(),
+            () -> {
+              for (int i = 0; i < unresolved.size(); i++) {
+                try {
+                  UUID uuid = unresolved.get(i);
+                  cache.put(
+                      uuid,
+                      new CachedPlayer(
+                          uuid, NameCacheUtil.resolveName(uuid), System.currentTimeMillis()));
+                } catch (IOException ignored) {
+                }
+              }
+              unresolved.clear();
+              try {
+                writeCacheToDisk();
+              } catch (IOException ignored) {
+              }
+            });
+  }
+
+  public static void writeCacheToDisk() throws IOException {
+    JsonWriter writer =
+        new JsonWriter(new FileWriter(new File(PGM.get().getDataFolder(), "uuidcache.json")));
+    writer.beginArray();
+    for (Entry<UUID, CachedPlayer> entry : cache.entrySet()) {
+      writer.beginObject();
+      writer.name("uuid").value(entry.getKey().toString());
+      writer.name("name").value(entry.getValue().getName());
+      writer.name("timestamp").value(entry.getValue().getTimestamp());
+      writer.endObject();
+    }
+    writer.endArray();
+    writer.close();
```

Comment:
When using GSON, you don't need to handle this yourself, you can just let gson do it for you. Also if you use try-with-resources syntax you don't need to close it yourself.

```java
        try (FileWriter writer = new FileWriter(new File(...))) {
            GSON.toJson(cache, writer);
        }
```